server/plugins/IOUGens.cpp

```diff
@@ -177,143 +287,111 @@ void Control_Ctor(Unit* unit) {
     }
 }
 
 void AudioControl_next_1(AudioControl* unit, int inNumSamples) {
     float** mapin = unit->mParent->mMapControls + unit->mSpecialIndex;
-    float* out = OUT(0);
-    int* mapRatep;
-    int mapRate;
-    float nextVal, curVal, valSlope;
-    float* prevVal;
-    prevVal = unit->prevVal;
-    curVal = prevVal[0];
-    mapRatep = unit->mParent->mControlRates + unit->mSpecialIndex;
-    mapRate = mapRatep[0];
-    World* world = unit->mWorld;
-    int32* touched = world->mAudioBusTouched;
-    int32 bufCounter = world->mBufCounter;
-    int32* channelOffsets = unit->mParent->mAudioBusOffsets;
+    int* mapRates = unit->mParent->mControlRates + unit->mSpecialIndex;
 
     if (*mapin != unit->m_prevBus) {
-        unit->m_busUsedInPrevCycle = false;
+        unit->m_busUsedInPrevCycle[0] = false;
         unit->m_prevBus = *mapin;
     }
 
-    switch (mapRate) {
-    case 0: {
-        for (int i = 0; i < inNumSamples; i++) {
-            out[i] = *mapin[0];
-        }
-    } break;
-    case 1: {
-        nextVal = *mapin[0];
-        valSlope = CALCSLOPE(nextVal, curVal);
-        for (int i = 0; i < inNumSamples; i++) {
-            out[i] = curVal;
-            curVal += valSlope;
-        }
-        unit->prevVal[0] = curVal;
-    } break;
-        /*
-         see case 2 comments above in definition for AudioControl_next_k
-         */
-    case 2: {
-        int thisChannelOffset = channelOffsets[unit->mSpecialIndex];
-        bool validOffset = thisChannelOffset >= 0;
-        int diff = bufCounter - touched[thisChannelOffset];
-        if (validOffset && diff == 0) {
-            Copy(inNumSamples, out, *mapin);
-            unit->m_busUsedInPrevCycle = true;
-        } else if (validOffset && diff == 1) {
-            if (unit->m_busUsedInPrevCycle) {
-                Fill(inNumSamples, out, 0.f);
-                unit->m_busUsedInPrevCycle = false;
-            } else
-                Copy(inNumSamples, out, *mapin);
-        } else {
-            Fill(inNumSamples, out, 0.f);
-            unit->m_busUsedInPrevCycle = false;
-        }
-    } break;
-    }
+    AudioControl_next_channel(unit, 0, mapin[0], mapRates[0], inNumSamples);
 }
 
 void AudioControl_Ctor(AudioControl* unit) {
-    unit->prevVal = (float*)RTAlloc(unit->mWorld, unit->mNumOutputs * sizeof(float));
-    unit->m_prevBus = NULL;
-    ClearUnitIfMemFailed(unit->prevVal);
-    for (int i = 0; i < unit->mNumOutputs; i++) {
-        unit->prevVal[i] = 0.0;
-    }
+    World* world = unit->mWorld;
+    int numChannels = unit->mNumOutputs;
+
+    unit->m_prevBus = nullptr;
+
+    size_t memSize = numChannels * (sizeof(float) + sizeof(bool));
+    char* mem = (char*)RTAlloc(world, memSize);
+    ClearUnitIfMemFailed(mem);
+
+    unit->m_prevVal = (float*)mem; // see AudioControl_Dtor()!
+    mem += numChannels * sizeof(float);
+    std::fill_n(unit->m_prevVal, numChannels, 0.f);
+
+    unit->m_busUsedInPrevCycle = (bool*)mem;
```

Comment:
are you sure this value gets freed in the destructor?

IIUC: You allocate mem via RTAlloct, use the first bits for `m_prevVal` and the 2nd section for `m_busUsedInPrevCycle`.

But then in the destructor you only seem to free `m_prevVal` which would not include the appended `numChannels * sizeof(bool)` section?
Or maybe I am not seeing something obvious here.

Reply:
We allocate a single chunk, so we must only free that chunk! `m_busUsedInPrevCycle` points into the memory, but it does not *own* it. `m_prevVal` actual points to the beginning of the memory chunk, that's why we pass it to `RT_Free` in the destructor.

`AudioControl_Dtor` actually has a comment (`// see AudioControl_Ctor()!`) but it puts too much exercise on the reader 😆 Will add a better comment!

Reply:
Side note: this is a common pattern in the Server code to reduce the number of (RT) memory allocations. It's essentially a crude linear allocator. In fact, Supernova has a dedicated helper class (`linear_allocator`) for this purpose.

Reply:
Ah sorry ... still try to wrap my head around C++ mem management. So`m_prevVal` takes the ownership of the memory address with its (implied) size and we simply cast a section of it to a float array pointer and cast the latter section to the bool array.

The allocator keeps track of memory address and its size and once we call the freeing of `m_prevVal` (whose address is equal to the initiated `mem`) it will free `m_prevVal` with the init range of `mem`, so including `m_prevVal` and `m_busUsedInPrevCycle`.

Makes sense, b/c you don't specify the range you want to free starting from a given pointer address...

Thanks for explaining this :)

Reply:
Side note: it's also important to first "allocate" the `float` array and only then the `bool` array, otherwise the `float` array might be misaligned.